atlasdb-commons/src/main/java/com/palantir/async/initializer/Callback.java

```diff
@@ -75,6 +75,24 @@ public void runWithRetry(R resource) {
         }
     }
 
+    /**
+     * Try init() once only, performing any necessary cleanup
+     */
+    public boolean runOnceOnly(R resource) {
+        try {
+            lock.lock();
```

Comment:
any chance we could get rid of the 'lock' and instead use synchronized here? In particular Java locks need to be used in the pattern:

```
lock.lock();
try {
    // do stuff
} finally {
    lock.unlock();
}
```

rather than the current one; this one is broken because the lock.lock() call may fail, and then you're unlocking a lock you never had :(